### PR TITLE
Cmake changes

### PR DESCRIPTION
**Pull request by Paul-Licameli**
The long promised/threatened changes to CMake scripts, to support more separation of modules and libraries from Audacity, and also to generate a graph of dependencies as a side effect at configuration time.

Leland especially, please review.  If satisfied, please rebase and merge.  Please don't squash, I'd rather keep the separated commits in the history.

Others are welcome to comment or ask questions.

### CODE REVIEW

**Review by crsib — CHANGES_REQUESTED** [review]
There are two minor issues, that should probably be fixed before we merge it

File: cmake-proxies/cmake-modules/AudacityFunctions.cmake [hunk 1]
```diff
@@ -204,3 +204,333 @@ macro( check_for_platform_version )
...
+   list( APPEND ${var}
+      PRIVATE
+         $<$<CXX_COMPILER_ID:MSVC>:/permissive->
+         $<$<CXX_COMPILER_ID:AppleClang,Clang>:-Wno-underaligned-exception-object>
+         $<$<CXX_COMPILER_ID:AppleClang,Clang>:-Werror=return-type>
+         $<$<CXX_COMPILER_ID:AppleClang,Clang>:-Werror=dangling-else>
+         $<$<CXX_COMPILER_ID:AppleClang,Clang>:-Werror=return-stack-address>
+	 # Yes, CMake will change -D to /D as needed for Windows:
+         -DWXINTL_NO_GETTEXT_MACRO
+         $<$<CXX_COMPILER_ID:MSVC>:/D _USE_MATH_DEFINES>
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply cmake-proxies/cmake-modules/AudacityFunctions.cmake:218]: I never had to do that to keep the GitHub builds working, but all right if you recommend it.

This one function is also a place where we can add more -Werror clang options, to turn certain warnings into build breakers and so enforce stricter requirements on code.  Such as the dangling-else ambiguity, a possible source of programmer errors which I disallowed earlier.

File: CMakeLists.txt [hunk 2]
```diff
@@ -426,13 +426,37 @@ include( "src/Experimental.cmake" )
...
+# Generate a picture of module dependencies
+string( JOIN "\n" GRAPH_EDGES ${GRAPH_EDGES} )
+# Choose edge attributes making it easy to hover at either end of edge
+# and see a tooltip describing the edge, in svg image
+file( WRITE modules.dot "digraph {
+   graph [rankdir=LR] edge [dir=both,arrowtail=inv] \n"
+   "${GRAPH_EDGES}"
+   "\n}\n"
+)
+execute_process( COMMAND
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt:457]: Unnecessary I think.

When the program named in execute_process does not exist, configuration still proceeds without an error.

Try it yourself:  change "dot" to "foobar" and configure the build.  It still works.

[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt:457]: In other words this useful graph is only an artifact of the configuration procedure, not a necessary step that the rest of configuration depends on.

**Comment by Paul-Licameli** [comment]
I have force-pushed onto more recent master.

I added /D NOMINMAX for Windows builds as Dmitry asked.

I changed the location of the generated .dot file as Leland asked.

I removed the commit for new file library/CMakeLists.txt so there is no conflict with Dmitry's telemetry branch.  I can make that change later.

I do not agree that it is necessary to remove the call to dot at configuration time.  If that program is not present, nothing happens and configuration still succeeds.  If it is present, it recreates a useful graph every time you reconfigure, with negligible cost.  I like having the graph done automatically in configuration, rather than as part of a build after configuration.

**Comment by lllucius** [comment]
Works for me.  I did notice the there wasn’t any problem when “dot” didn’t exist.  I just didn’t know if it was a “time consuming” process if id DID exist.  I guess I could have installed graphviz to find out. 😊

**Comment by Paul-Licameli** [comment]
Thanks, but sorry, I just did the NOMINMAX part wrong and now Windows build is broken.  I'll fix that.

**Comment by Paul-Licameli** [comment]
Hey wait, did you amend the commit for me?  I'm confused how that happened.

**Comment by Paul-Licameli** [comment]
Oh, I see I force-pushed the corrected branch, and then you merged it while I was waiting for the build to finish!  So all's well that ends well.

**Pull request merged**